README.md
[![Python interface to WAVEWATCH III data][logo]][github-link]


# WAVEWATCH III data in Python

[![Documentation status][rtd-badge]][rtd-link]
[![Code-style: Black][black-badge]][black-link]
[![Testing status][testing-badge]][testing-link]
![Flake8][flake8-badge]


## About


<!-- start-abstract -->

The *bmi_wavewatch3* Python package provides both a command line interface and
a programming interface for downloading and working with [WAVEWATCH III] data.

*bmi_wavewatch3* provides access to the following raster data sources,

- 30 year wave hindcast [Phase 1]
- 30 year wave hindcast [Phase 2]
- Production hindcast [Singlegrid]
- Production hindcast [Multigrid]

All data sources provide both global and regional grids.

[phase 1]: https://polar.ncep.noaa.gov/waves/hindcasts/nopp-phase1.php
[phase 2]: https://polar.ncep.noaa.gov/waves/hindcasts/nopp-phase2.php
[singlegrid]: https://polar.ncep.noaa.gov/waves/hindcasts/prod-nww3.php
[multigrid]: https://polar.ncep.noaa.gov/waves/hindcasts/prod-multi_1.php

<!-- end-abstract -->

[**See the bmi-wavewatch3 documentation for more information**](https://bmi-wavewatch3.readthedocs.io/en/latest/).

## Installation

To install the latest release of *bmi-wavewatch3* using *pip*, simply run the following
in your terminal of choice:

```bash
pip install bmi-wavewatch3
```

For a full description of how to install *bmi-wavewatch3*, including using *mamba*/*conda*,
please see the documentation for our [installation instructions].

## Source code

If you would like to modify or contribute code to *bmi-wavewatch3* or use the very
latest development version, please see the documentation that describes how to
[install bmi-wavewatch3 from source].


[install bmi-wavewatch3 from source]: https://bmi-wavewatch3.readthedocs.io/en/master/install/developer_install.html
[installation instructions]: https://bmi-wavewatch3.readthedocs.io/en/master/installation.html
[multigrid data]: https://polar.ncep.noaa.gov/waves/hindcasts/multi_1/
[singlegrid data]: https://polar.ncep.noaa.gov/waves/hindcasts/nww3/
[wavewatch iii]: https://polar.ncep.noaa.gov/waves
[wavewatch iii description]: https://polar.ncep.noaa.gov/waves/wavewatch/
[wavewatch iii hindcasts]: http://polar.ncep.noaa.gov/waves/hindcasts/
[wavewatch iii thredds]: https://www.ncei.noaa.gov/thredds-ocean/catalog/ncep/nww3/catalog.html


[black-badge]: https://github.com/csdms/bmi-wavewatch3/workflows/Black/badge.svg
[black-link]: https://github.com/ambv/black
[flake8-badge]: https://github.com/csdms/bmi-wavewatch3/workflows/Flake8/badge.svg
[github-link]: https://github.com/csdms/bmi-wavewatch3
[logo]: https://github.com/csdms/bmi-wavewatch3/raw/main/docs/source/_static/bmi-wavewatch3-logo-light.svg
[rtd-badge]: https://readthedocs.org/projects/bmi-wavewatch3/badge/?version=latest
[rtd-link]: https://bmi-wavewatch3.readthedocs.io/en/latest/?badge=latest
[testing-badge]: https://github.com/csdms/bmi-wavewatch3/actions/workflows/test.yml/badge.svg
[testing-link]: https://github.com/csdms/bmi-wavewatch3/actions/workflows/test.yml
[ww3-global-image]: https://raw.githubusercontent.com/csdms/bmi-wavewatch3/main/docs/source/_static/ww3_global_swh-light.png


CHANGES.md
# Release Notes


```{towncrier-draft-entries} Not yet released
```

<!-- towncrier-draft-entries:: Not yet released -->

<!-- towncrier release notes start -->

## 0.2.0 (2022-06-17)

### New Features

- Added a new subcommand, *plot*, to the *ww3* command-line program.
  `ww3 plot` with download (if the data files are not already cached) and
  create a plot of the requested data. ([#13](https://github.com/csdms/bmi-wavewatch3/issues/13))

### Bug Fixes

- Fixed a bug in the reporting of an error caused by an invalide datatime
  string. ([#13](https://github.com/csdms/bmi-wavewatch3/issues/13))

## 0.1.1 (2022-06-10)

### Other Changes and Additions

- Set up GitHub Action to create a source distribution and push it to
  *TestPyPI*. This action is only run if the version tag is a prerelease version
  (i.e. the version string ends with `[ab][0-9]+`). ([#10](https://github.com/csdms/bmi-wavewatch3/issues/10))
- Set up GitHub Action to create a source distribution and push it to
  *PyPI*. This action is only run if the version tag is a release version
  (i.e. the version string doesn't end with `[ab][0-9]+`). ([#11](https://github.com/csdms/bmi-wavewatch3/issues/11))

## 0.1.1b1 (2022-06-09)

### New Features

- Added `ww3` command line interface to download WaveWatch III data by date,
  region and quantity (significant wave height, wind speed, etc.). ([#1](https://github.com/csdms/bmi-wavewatch3/issues/1))
- Added `WaveWatch3` class, which is the main access point for users of this package.
  This class downloads WaveWatch III data files (if not already cached) and provides a
  view of the data as an xarray Dataset. Users can then advance through the data
  month-by-month, downloading additional data as necessary. ([#3](https://github.com/csdms/bmi-wavewatch3/issues/3))
- Added the `ww3 clean` subcommand that removes cached data files. ([#4](https://github.com/csdms/bmi-wavewatch3/issues/4))
- Added `BMIWaveWatch3` class to provide a Basic Model Interface for the
  *wavewatch3* package. ([#5](https://github.com/csdms/bmi-wavewatch3/issues/5))
- Added additional WaveWatch III data sources from which users can fraw data
  from. ([#6](https://github.com/csdms/bmi-wavewatch3/issues/6))
- Added `fetch` method to WaveWatch3 to mimic the command line program
  `ww3 fetch`. ([#7](https://github.com/csdms/bmi-wavewatch3/issues/7))
- Added additional data sources from which to retreive data from. Available
  data sources now include: Phase 1, Phase 2, Multigrid, Multigrid-extended,
  and Multigrid-thredds. ([#7](https://github.com/csdms/bmi-wavewatch3/issues/7))
- Added `ww3 info` command to print information (e.g. available grids, quantities,
  etc.) about data sources. ([#7](https://github.com/csdms/bmi-wavewatch3/issues/7))
- Added a `step` property to `WaveWatch3` to track the current time slice
  of the data cube. This property is also settable so that a user can use it to
  advance throught the data (additional data are downloaded in the background as
  needed). ([#8](https://github.com/csdms/bmi-wavewatch3/issues/8))
- Dates can now be specified as iso-formatted date/time strings. For example,
  "1944-06-06T06:30". ([#8](https://github.com/csdms/bmi-wavewatch3/issues/8))
- Rename package to `bmi_wavewatch3`. This follows the convention used by other
  CSDMS data components. ([#9](https://github.com/csdms/bmi-wavewatch3/issues/9))

### Documentation Enhancements

- Added package description, installation, usage, and an example to the
  documentation. ([#8](https://github.com/csdms/bmi-wavewatch3/issues/8))

### Other Changes and Additions

- Set up continuous integration using GitHub actions. This includes tests to
  ensure that the code is styled according to *black*, is free of lint, and
  passes all unit tests. ([#2](https://github.com/csdms/bmi-wavewatch3/issues/2))
- Added more unit tests, particularly for data sources. ([#7](https://github.com/csdms/bmi-wavewatch3/issues/7))
- Added a GitHub action to build the sphinx-based documentation as part of the
  continuous integration. ([#8](https://github.com/csdms/bmi-wavewatch3/issues/8))
- Better error reporting for the command line interface for HTTP errors when
  retreiving data as well as input validation. ([#8](https://github.com/csdms/bmi-wavewatch3/issues/8))
- Set up GitHub Action to create a source distribution and push it to
  *TestPyPI*. This action is only run if the version tag is a prerelease version
  (i.e. the version string ends with `[ab][0-9]+`). ([#10](https://github.com/csdms/bmi-wavewatch3/issues/10))


wavewatch3/source.py
import datetime
import pathlib
import urllib

from .errors import ChoiceError, DateValueError


class _WaveWatch3Source:
    SCHEME = "https"
    NETLOC = ""
    PREFIX = ""

    GRIDS = {
        "glo_30m",
        "ao_30m",
        "at_10m",
        "wc_10m",
        "ep_10m",
        "ak_10m",
        "at_4m",
        "wc_4m",
        "ak_4m",
    }
    QUANTITIES = {"wind", "hs", "tp", "dp"}
    MIN_DATE = None
    MAX_DATE = None

    def __init__(self, date, quantity, grid="glo_30m"):
        self._date = datetime.datetime.fromisoformat(self.validate_date(date))
        self._quantity = self.validate_quantity(quantity)
        self._grid = self.validate_grid(grid)

    @property
    def path(self):
        raise NotImplementedError("path")

    @property
    def filename(self):
        raise NotImplementedError("filename")

    def __str__(self):
        return urllib.parse.urlunparse(
            [
                self.SCHEME,
                self.NETLOC,
                str(self.path),
                "",
                "",
                "",
            ]
        )

    def __repr__(self):
        date = self._date.isoformat()
        return f"{self.__class__.__module__}.{self.__class__.__name__}({date!r}, {self.quantity!r}, grid={self.grid!r})"

    def __eq__(self, other):
        return str(self) == str(other)

    @property
    def grid(self):
        return self._grid

    @grid.setter
    def grid(self, grid):
        self.validate_grid(grid)
        self._grid = grid

    @property
    def date(self):
        return self._date.isoformat(timespec="hours")

    @date.setter
    def date(self, date):
        self._date = datetime.datetime.fromisoformat(self.validate_date(date))

    @property
    def quantity(self):
        return self._quantity

    @quantity.setter
    def quantity(self, quantity):
        self._quantity = self.validate_quantity(quantity)

    @property
    def year(self):
        return self._date.year

    @year.setter
    def year(self, year):
        try:
            self._date = datetime.date(year, self.month, self._date.day)
        except ValueError as error:
            raise DateValueError(str(error))

    @property
    def month(self):
        return self._date.month

    @month.setter
    def month(self, month):
        try:
            self._date = datetime.datetime(self.year, month, self._date.day)
        except ValueError as error:
            raise DateValueError(str(error))

    @classmethod
    def validate_quantity(cls, quantity):
        if quantity not in cls.QUANTITIES:
            raise ChoiceError(quantity, cls.QUANTITIES)
        return quantity

    @classmethod
    def validate_grid(cls, grid):
        if grid not in cls.GRIDS:
            raise ChoiceError(grid, cls.GRIDS)
        return grid

    @classmethod
    def validate_date(cls, date):
        try:
            datetime.datetime.fromisoformat(date)
        except ValueError as error:
            raise DateValueError(error)
        date_in_range_or_raise(date, lower=cls.MIN_DATE, upper=cls.MAX_DATE)

        return date


def date_in_range_or_raise(date, lower=None, upper=None):
    date = datetime.datetime.fromisoformat(date)

    lower = datetime.datetime.fromisoformat(lower) if lower else lower
    if lower and date < lower:
        raise DateValueError(
            f"{date}: date is less than minimum date for this dataset ({lower})"
        )

    upper = datetime.datetime.fromisoformat(upper) if upper else upper
    if upper and date > upper:
        raise DateValueError(
            f"{date}: date is greater than maximum date for this dataset ({upper})"
        )


class WaveWatch3SourcePhase1(_WaveWatch3Source):
    """
    https://polar.ncep.noaa.gov/waves/hindcasts/nopp-phase1.php
    """

    SCHEME = "https"
    NETLOC = "polar.ncep.noaa.gov"
    PREFIX = "/waves/hindcasts/nopp-phase1"

    GRIDS = {
        "ak_4m",
        "ak_10m",
        "ecg_4m",
        "ecg_10m",
        # "glo_30m_ext",
        "glo_30m",
        "med_10m",
        "nsb_4m",
        "nsb_10m",
        "nwio_10m",
        "oz_4m",
        "oz_10m",
        "pi_10m",
        "wc_4m",
        "wc_10m",
    }

    MIN_DATE = "1979-01-01"
    MAX_DATE = "2009-12-31"

    def __init__(self, date, quantity, grid="glo_30m"):
        super().__init__(date, quantity, grid=grid)

    @property
    def path(self):
        path = (
            self.PREFIX / pathlib.PurePosixPath(f"{self.year}{self.month:02d}") / "grib"
        )
        return path / self.filename

    @property
    def filename(self):
        return f"multi_reanal.{self.grid}.{self.quantity}.{self.year}{self.month:02d}.grb2.gz"


class WaveWatch3SourcePhase2(_WaveWatch3Source):
    """
    https://polar.ncep.noaa.gov/waves/hindcasts/nopp-phase2.php
    """

    SCHEME = "https"
    NETLOC = "polar.ncep.noaa.gov"
    PREFIX = "/waves/hindcasts/nopp-phase2"

    GRIDS = {
        "ak_4m",
        "ak_10m",
        "ecg_4m",
        "ecg_10m",
        "glo_30m_ext",
        "med_10m",
        "nsb_4m",
        "nsb_10m",
        "nwio_10m",
        "oz_4m",
        "oz_10m",
        "pi_10m",
        "wc_4m",
        "wc_10m",
    }

    MIN_DATE = "1979-01-01"
    MAX_DATE = "2009-12-31"

    def __init__(self, date, quantity, grid="glo_30m_ext"):
        super().__init__(date, quantity, grid=grid)

    @property
    def path(self):
        path = (
            self.PREFIX
            / pathlib.PurePosixPath(f"{self.year}{self.month:02d}")
            / "gribs"
        )
        return path / self.filename

    @property
    def filename(self):
        return (
            f"multi_reanal.{self.grid}.{self.quantity}.{self.year}{self.month:02d}.grb2"
        )


class WaveWatch3SourceMultigrid(_WaveWatch3Source):
    """
    https://polar.ncep.noaa.gov/waves/hindcasts/prod-multi_1.php
    """

    SCHEME = "https"
    NETLOC = "polar.ncep.noaa.gov"
    PREFIX = "/waves/hindcasts/multi_1"

    MIN_DATE = "2005-02-01"
    MAX_DATE = "2019-05-31"

    @property
    def path(self):
        path = (
            self.PREFIX
            / pathlib.PurePosixPath(f"{self.year}{self.month:02d}")
            / "gribs"
        )
        return path / self.filename

    @property
    def filename(self):
        return f"multi_1.{self.grid}.{self.quantity}.{self.year}{self.month:02d}.grb2"


class WaveWatch3SourceMultigridExt(_WaveWatch3Source):
    """
    https://polar.ncep.noaa.gov/waves/hindcasts/prod-multi_1.php
    """

    SCHEME = "https"
    NETLOC = "polar.ncep.noaa.gov"
    PREFIX = "/waves/hindcasts/multi_1"

    GRIDS = {
        "glo_30m",
        "ao_30m",
        "at_10m",
        "wc_10m",
        "ep_10m",
        "ak_10m",
        "at_4m",
        "wc_4m",
        "ak_4m",
    }
    QUANTITIES = {"wind", "hs", "tp", "dp", "phs", "ptp", "pdir"}  # added 2017 onward

    MIN_DATE = "2017-02-01"
    MAX_DATE = "2019-05-31"

    @property
    def path(self):
        path = (
            self.PREFIX
            / pathlib.PurePosixPath(f"{self.year}{self.month:02d}")
            / "gribs"
        )
        return path / self.filename

    @property
    def filename(self):
        return f"multi_1.{self.grid}.{self.quantity}.{self.year}{self.month:02d}.grb2"


class WaveWatch3SourceThredds(_WaveWatch3Source):
    SCHEME = "https"
    NETLOC = "www.ncei.noaa.gov"
    PREFIX = "/thredds-ocean/fileServer/ncep/nww3"

    MIN_DATE = "2005-02-01"
    MAX_DATE = "2019-05-31"

    @property
    def path(self):
        path = self.PREFIX / pathlib.PurePosixPath(f"{self.year}", f"{self.month:02d}")
        path /= self.grid if self.year < 2017 else "gribs"
        return path / self.filename

    @property
    def filename(self):
        return f"multi_1.{self.grid}.{self.quantity}.{self.year}{self.month:02d}.grb2"


class WaveWatch3SourceSinglegrid(_WaveWatch3Source):
    """
    https://polar.ncep.noaa.gov/waves/hindcasts/prod-nww3.php
    """

    SCHEME = "https"
    NETLOC = "polar.ncep.noaa.gov"
    PREFIX = "/waves/hindcasts/nww3"

    GRIDS = {"akw", "enp", "nah", "nph", "nww3", "wna"}

    MIN_DATE = "1999-07-01"
    MAX_DATE = "2006-09-30"

    def __init__(self, date, quantity, grid="nww3"):
        super().__init__(date, quantity, grid=grid)

    @property
    def path(self):
        return pathlib.PurePosixPath(self.PREFIX) / self.filename

    @property
    def filename(self):
        return f"{self.grid}.{self.quantity}.{self.year}{self.month:02d}.grb"


SOURCES = {
    "multigrid": WaveWatch3SourceMultigrid,
    "multigrid-extended": WaveWatch3SourceMultigridExt,
    "multigrid-thredds": WaveWatch3SourceThredds,
    "phase1": WaveWatch3SourcePhase1,
    "phase2": WaveWatch3SourcePhase2,
    "singlegrid": WaveWatch3SourceSinglegrid,
}
